introducing-pgzx.mdx

```diff
@@ -0,0 +1,208 @@
+```
+
+It points exactly where the error happened. This happens because Zig has runtime checks depending on the [build mode](https://ziglang.org/documentation/master/#Build-Mode). The `ReleaseSafe` mode, for example, trades a bit of performance for more safety checks.
+
+It's worth pointing out that the stacktrace worked so well because the error was in Zig code. When building Postgres extensions, you often have to call Postgres APIs and you can still cause segmentation faults if you use them incorrectly.
+
+### Memory management
+
+Postgres uses a [memory context system](https://github.com/postgres/postgres/blob/master/src/backend/utils/mmgr/README) to manage memory. Memory allocated in a "context" can be freed all at once (for example, when a query execution is finished), which simplifies memory management significantly, because you only need to track contexts, not individual allocations. Contexts are also hierarchical, so you can create a context that is a child of another context, and when the parent context is freed, all children are freed as well.
+
+Another advantage of memory contexts is that they improve memory monitoring, because context have names and you can see how much memory is used by each context. This is useful for debugging memory leaks, for example.
+
+A nice thing about the Zig standard library is that all functions/objects that allocate memory receive an allocator as a parameter. This makes allocations [more explicit](https://notes.eatonphil.com/2024-03-15-zig-rust-and-other-languages.html) but also makes it easy to use a custom allocator. pgzx defines custom allocators that wrap the Postgres memory contexts and make them available to Zig code.
+
+Here is an example that creates a new context as a child of the current context and gets the allocator for it:
+
+```
+var memctx = try pgzx.mem.createAllocSetContext("zig_context", .{ .parent = pg.CurrentMemoryContext });
+const allocator = memctx.allocator();
+```
+
+### Error handling
+
+Another Postgres API that you will very likely need to know about in a more complex extension is the error handling. Postgres implements "exceptions" in C via `setjmp/longjmp` and provides a set of macros to throw and catch them ([PG_TRY/PG_CATCH](https://github.com/postgres/postgres/blob/master/src/include/utils/elog.h#L318)).
+
+The issue is that long jumps could bypass the Zig control flow, for example `errdefer` blocks might not be executed. This means that if your extension calls into Postgres APIs, and those APIs can throw errors, long jumps might skip your `defer` and `errdefer` blocks!
+
+Luckily, pgzx is here to help. It provides a set of functions that allow you to catch the Postgres exceptions and convert them into Zig errors. For example:
+
+```
+var errctx = pgzx.err.Context.init();
+defer errctx.deinit();
+if (errctx.pg_try()) {
+    // zig code that calls several Postgres C functions.
+} else {
+    return errctx.errorValue();
+}
+```
+
```

Comment:
We also provide `pgzx.err.wrap( ... )` helper that implements that pattern for you. Use it to wrap a single function call and turn a Postgres exception into a Zig error that you can rethrow later if you want.

Eventually we should come up with some guidelines about error handling. I haven't made my mind yet. Right now in 'library' like code I make sure to return Zig errors and convert postgres exceptions into Zig errors. This helps users of my lib to not have to think about capturing postgres errros. But honestly, due to memory manamgent and error handling working hand in hand in Postgres. One can often just rely on the Postgres exceptions and only capture the errors if some cleanup is required... E.g. when just implementing "functions" that do not acquire external resources Postgres error handling is all you need.

Maybe time will tell which error handling strategy is a better fit for which kind of extensions.

Reply:
I kept the example code blocks relatively short, and only gave one version for each. I could replace this one with the  `wrap()` version, though.

Reply:
Let's stick with the more verbose example.